python/ml4ir/applications/ranking/t_test.py

```diff
@@ -1,5 +1,324 @@
+    non-centrality parameter.
+    :py:func:`scipy.optimize.brenth` is used to solve power equations for other variables (i.e.
+    sample size, effect size, or significance level). If the solving fails, a nan value is
+    returned.
+    Results have been tested against GPower and the
+    `pwr <https://cran.r-project.org/web/packages/pwr/pwr.pdf>`_ R package.
+    Examples
+    --------
+    1. Compute power of a one-sample T-test given ``d``, ``n`` and ``alpha``
+    >>> from pingouin import power_ttest
+    >>> print('power: %.4f' % power_ttest(d=0.5, n=20, contrast='one-sample'))
+    power: 0.5645
+    2. Compute required sample size given ``d``, ``power`` and ``alpha``
+    >>> print('n: %.4f' % power_ttest(d=0.5, power=0.80, alternative='greater'))
+    n: 50.1508
+    3. Compute achieved ``d`` given ``n``, ``power`` and ``alpha`` level
+    >>> print('d: %.4f' % power_ttest(n=20, power=0.80, alpha=0.05, contrast='paired'))
+    d: 0.6604
+    4. Compute achieved alpha level given ``d``, ``n`` and ``power``
+    >>> print('alpha: %.4f' % power_ttest(d=0.5, n=20, power=0.80, alpha=None))
+    alpha: 0.4430
+    5. One-sided tests
+    >>> from pingouin import power_ttest
+    >>> print('power: %.4f' % power_ttest(d=0.5, n=20, alternative='greater'))
+    power: 0.4634
+    >>> print('power: %.4f' % power_ttest(d=0.5, n=20, alternative='less'))
+    power: 0.0007
+    """
+    # Check the number of arguments that are None
+    n_none = sum([v is None for v in [d, n, power, alpha]])
+    if n_none != 1:
+        raise ValueError("Exactly one of n, d, power, and alpha must be None.")
+
+    # Safety checks
+    assert alternative in [
+        "two-sided",
+        "greater",
+        "less",
+    ], "Alternative must be one of 'two-sided' (default), 'greater' or 'less'."
+    assert contrast.lower() in ["one-sample", "paired", "two-samples"]
+    tsample = 2 if contrast.lower() == "two-samples" else 1
+    tside = 2 if alternative == "two-sided" else 1
+    if d is not None and tside == 2:
+        d = abs(d)
+    if alpha is not None:
+        assert 0 < alpha <= 1
+    if power is not None:
+        assert 0 < power <= 1
+
+    if alternative == "less":
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(alpha / tside, dof)
+            return stats.nct.cdf(tcrit, dof, nc)
+
+    elif alternative == "two-sided":
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(1 - alpha / tside, dof)
+            return stats.nct.sf(tcrit, dof, nc) + stats.nct.cdf(-tcrit, dof, nc)
+
+    else:  # Alternative = 'greater'
+
+        def func(d, n, power, alpha):
+            dof = (n - 1) * tsample
+            nc = d * np.sqrt(n / tsample)
+            tcrit = stats.t.ppf(1 - alpha / tside, dof)
+            return stats.nct.sf(tcrit, dof, nc)
```

Comment:
Why can't `alternative` just be an argument and then you define the function if `if-else` blocks? This seems like a lot of code duplication and I am not sure it is a good coding pattern.

Reply:
Perhaps. However, I extracted the code from the lib to not add dependency and I didn't spend time to improve it.